### PR TITLE
dry-run-deploys tool

### PR DESCRIPTION
**Pull request by dwerner**
Resolves issues: #2097, #2121 

Brings in dry-run-deploys cli tool.

Replaces and combines old PRs #2168 and #2102

### CODE REVIEW

File: execution_engine/src/storage/trie_store/lmdb.rs [hunk 1]
```diff
@@ -39,7 +39,7 @@
 //! let tmp_dir = tempdir().unwrap();
 //! let map_size = 4096 * 2560;  // map size should be a multiple of OS page size
 //! let max_readers = 512;
-//! let env = LmdbEnvironment::new(&tmp_dir.path().to_path_buf(), map_size, max_readers, true).unwrap();
+//! let env = LmdbEnvironment::new(&tmp_dir.path().to_path_buf(), map_size, max_readers, true, 2).unwrap();
```
Comment (xcthulhu) [inline execution_engine/src/storage/trie_store/lmdb.rs]:
```suggestion
//! let env = LmdbEnvironment::new(&tmp_dir.path().to_path_buf(), map_size, max_readers, true).unwrap();
```

File: node/src/components/consensus/validator_change.rs [hunk 2]
```diff
@@ -142,9 +142,9 @@ mod tests {
 
         let mut era1_metadata = era0_metadata.clone();
         let added_validator = PublicKey::random(&mut rng);
+        let expected_change = vec![(added_validator.clone(), ValidatorChange::Added)];
```
Comment (xcthulhu) [inline node/src/components/consensus/validator_change.rs:145]: Note: needed for clippy

File: node/src/types/block.rs [hunk 3]
```diff
@@ -90,10 +90,10 @@ static ERA_END: Lazy<EraEnd> = Lazy::new(|| {
     EraEnd::new(era_report, next_era_validator_weights)
 });
 static FINALIZED_BLOCK: Lazy<FinalizedBlock> = Lazy::new(|| {
-    let deploy_hashes = vec![*Deploy::doc_example().id()];
+    let transfer_hashes = vec![*Deploy::doc_example().id()];
```
Comment (xcthulhu) [inline node/src/types/block.rs:93]: good catch

File: types/src/bytesrepr.rs [hunk 4]
```diff
@@ -21,6 +21,9 @@ use num_integer::Integer;
 use num_rational::Ratio;
 use serde::{Deserialize, Serialize};
 
+#[cfg(feature = "std")]
+use thiserror::Error;
+
```
Comment (xcthulhu) [inline types/src/bytesrepr.rs]:
```suggestion
```

File: types/src/bytesrepr.rs [hunk 5]
```diff
@@ -100,6 +100,7 @@ pub fn allocate_buffer<T: ToBytes>(to_be_serialized: &T) -> Result<Vec<u8>, Erro
 
 /// Serialization and deserialization errors.
 #[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
```
Comment (xcthulhu) [inline types/src/bytesrepr.rs:103]: Note: needed for clients who use this library with `std` (for instance, this enables using `eyre` and `anyhow`)

File: utils/retrieve-state/src/storage.rs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        Ok(path.into())
+    } else {
+        Ok(env::current_dir()?.join(path))
+    }
+}
+
+pub fn create_storage(chain_download_path: impl AsRef<Path>) -> Result<Storage, anyhow::Error> {
+    let chain_download_path = normalize_path(chain_download_path)?;
+    let mut storage_config = StorageConfig::default();
+    storage_config.path = chain_download_path.clone();
```
Comment (xcthulhu) [inline utils/retrieve-state/src/storage.rs:131]: 🤯

File: utils/retrieve-state/src/storage.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    Ok(Storage::new(
+        &WithDir::new(chain_download_path, storage_config),
+        None,
+        ProtocolVersion::from_parts(1, 0, 0),
```
Comment (xcthulhu) [inline utils/retrieve-state/src/storage.rs]:
```suggestion
        ProtocolVersion::from_parts(0, 0, 0),
```

**Review by xcthulhu — APPROVED** [review]
LGTM

File: node/src/components.rs [hunk 8]
```diff
@@ -21,7 +21,7 @@ pub mod in_memory_network;
 pub(crate) mod metrics;
 pub(crate) mod networking_metrics;
 pub(crate) mod small_network;
-pub(crate) mod storage;
+pub mod storage;
```
Comment (xcthulhu) [inline node/src/components.rs:24]: Would be nice if this was `pub(crate)`, so we should consider factoring out storage into another crate.

**Comment by dwerner** [comment]
bors try

**Pull request merged**